internal/evmreader/evmreader.go

```diff
@@ -0,0 +1,344 @@
+		case <-headers:
+
+			err = r.checkForNewInputs(ctx)
+			if err != nil {
+				slog.Error("Error checking got new inputs",
+					"error",
+					err,
+				)
+			}
+
+		}
+	}
+}
+
+// Check if is there new Inputs for all running Applications
+func (r *EvmReader) checkForNewInputs(ctx Context) error {
+
+	// Get All Applications
+	apps, err := r.repository.GetAllRunningApplications(ctx)
+	if err != nil {
+		return err
+	}
+
+	groupedApps := r.classifyApplicationsByLastProcessedInput(apps)
+
+	for lastProcessedBlock, apps := range groupedApps {
+
+		// Safeguard: Only check blocks starting from the block where the InputBox
+		// contract was deployed as Inputs can be added to that same block
+		if lastProcessedBlock < r.config.inputBoxBlockNumber {
+			lastProcessedBlock = r.config.inputBoxBlockNumber - 1
+		}
+
+		currentMostRecentFinalizedHeader, err := r.fetchMostRecentHeader(
+			ctx,
+			r.config.defaultBlock,
+		)
+		if err != nil {
+			slog.Error("Error fetching most recent block",
+				"last default block",
+				r.config.defaultBlock,
+				"error",
+				err)
+			continue
+		}
+		currentMostRecentFinalizedBlockNumber := currentMostRecentFinalizedHeader.Number.Uint64()
+
+		if currentMostRecentFinalizedBlockNumber > lastProcessedBlock {
+
+			err = r.readInputs(ctx,
+				lastProcessedBlock+1,
+				currentMostRecentFinalizedBlockNumber,
+				apps,
+			)
+			if err != nil {
+				slog.Error("Error reading inputs",
+					"start",
+					lastProcessedBlock+1,
+					"end",
+					currentMostRecentFinalizedBlockNumber,
+					"error",
+					err)
+				continue
+			}
+		} else if lastProcessedBlock < currentMostRecentFinalizedBlockNumber {
+			slog.Warn(
+				"current most recent block is lower than the last processed one",
+				"most recent block",
+				currentMostRecentFinalizedBlockNumber,
+				"last processed",
+				lastProcessedBlock,
+			)
+		}
+	}
+
+	return nil
+}
+
+// Group Applications that have processed til the same block height
+func (r *EvmReader) classifyApplicationsByLastProcessedInput(
+	apps []Application,
+) map[uint64][]Application {
+	result := make(map[uint64][]Application)
+	for _, app := range apps {
+		result[app.LastProcessedBlock] = append(result[app.LastProcessedBlock], app)
+	}
+
+	return result
+}
+
+// Fetch the most recent header up till the
+// given default block
+func (r *EvmReader) fetchMostRecentHeader(
+	ctx Context,
+	defaultBlock DefaultBlock,
+) (*types.Header, error) {
+
+	var defaultBlockNumber int64
+	switch defaultBlock {
+	case model.DefaultBlockStatusPending:
+		defaultBlockNumber = rpc.PendingBlockNumber.Int64()
+	case model.DefaultBlockStatusLatest:
+		defaultBlockNumber = rpc.LatestBlockNumber.Int64()
+	case model.DefaultBlockStatusFinalized:
+		defaultBlockNumber = rpc.FinalizedBlockNumber.Int64()
+	case model.DefaultBlockStatusSafe:
+		defaultBlockNumber = rpc.SafeBlockNumber.Int64()
+	default:
+		return nil, fmt.Errorf("Default block '%v' not supported", defaultBlock)
+	}
```

Comment:
Since the `defaultBlock` is an enum, is there any chance we fall on this case?

Reply:
If a new value is added to the enum, but this switch case is not updated